### ISSUE TITLE
Memory bug on the web

### ISSUE THREAD

**Issue by filiph** [flutter/flutter#153678]
### Steps to reproduce

1. Create new Flutter project: `flutter create bug && cd bug`
2. Add a dependency on `pkg:vector_math`: `flutter pub add vector_math` (dependency on `vector_math` is almost certainly not needed but it makes the repro code cleaner)
3. Replace `lib/main.dart` with the code below (in code sample section).
4. Run in any non-web target (e.g. `flutter run -d macos` or `flutter build macos`) to ensure there is no memory leak (memory stays at a certain level):

![Screenshot 2024-08-18 at 12 17 38](https://github.com/user-attachments/assets/4b578404-a389-47f4-bb52-ebb5565ab563)

6. Run on the web (e.g. `flutter build web`) in any configuration (release / profile / debug, wasm or not, etc.)

### Expected results

Used memory doesn't grow.

### Actual results

Memory used by the web app goes up linearly at several megabytes per second:

![Screenshot 2024-08-18 at 12 57 41](https://github.com/user-attachments/assets/277e8b05-f28d-4232-8d21-a323547d3903)

After a couple of minutes, the memory consumed reaches 1.0 GB and keeps growing:

![Screenshot 2024-08-18 at 12 22 33](https://github.com/user-attachments/assets/f7e23ae1-87a5-4c86-9c58-7a6c313acec2)

I tried debugging this through Dart DevTools but Dart DevTools don't have access to memory info for web apps. All I can do is use Chrome DevTools. I can't say I understand what I'm looking at in Chrome DevTools, but here are some things I noticed:

1. Basically all the RAM is taken by some `JSArrayBufferData` object:

![Screenshot 2024-08-18 at 13 03 55](https://github.com/user-attachments/assets/15d93068-56c0-43d4-a085-8a7f35c478cb)

2. This object seems to have something in common with a `flutterCanvasKit` object, and specifically with something called `globalThis`.

![Screenshot 2024-08-19 at 09 31 43](https://github.com/user-attachments/assets/0162cd16-f2d4-42b4-92e7-e9025f6682e3)

Snapshot file from Chrome DevTools: [Heap-20240818T130707.heapsnapshot.zip](https://github.com/user-attachments/files/16657126/Heap-20240818T130707.heapsnapshot.zip)

I initially thought this had something to do with Flame but as you can see from the sample, the same issue appears on vanilla Flutter as well. // cc @zoeyfan 

### Code sample

<details open><summary>Code sample</summary>

It's possible that the code could be made simpler but before I invest that kind of work, I'd like to make sure this is actually a bug and that I'm not overlooking something obvious or doing something stupid.

```dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import 'package:vector_math/vector_math_64.dart' hide Colors;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  const MyHomePage({super.key, required this.title});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(title),
      ),
      body: const Center(
        child: MemoryPressureWidget(),
      ),
    );
  }
}

class MemoryPressureWidget extends StatefulWidget {
  const MemoryPressureWidget({super.key});

  @override
  State<MemoryPressureWidget> createState() => _MemoryPressureWidgetState();
}

class _MemoryPressureWidgetState extends State<MemoryPressureWidget>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  final List<PairedWanderer> wanderers = [];

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        for (final wanderer in wanderers)
          PairedWandererWidget(wanderer: wanderer),
      ],
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  void initState() {
    super.initState();
    _createBatch(1000, const Size(500, 500));

    _controller = AnimationController(vsync: this);
  }

  void _createBatch(int batchSize, Size worldSize) {
    assert(batchSize.isEven);
    final random = Random(42);
    for (var i = 0; i < batchSize / 2; i++) {
      final a = PairedWanderer(
        velocity: (Vector2.random() - Vector2.all(0.5))..scale(100),
        worldSize: worldSize,
        position: Vector2(worldSize.width * random.nextDouble(),
            worldSize.height * random.nextDouble()),
      );
      final b = PairedWanderer(
        velocity: (Vector2.random() - Vector2.all(0.5))..scale(100),
        worldSize: worldSize,
        position: Vector2(worldSize.width * random.nextDouble(),
            worldSize.height * random.nextDouble()),
      );
      a.otherWanderer = b;
      b.otherWanderer = a;
      wanderers.add(a);
      wanderers.add(b);
    }
  }
}

class PairedWanderer {
  PairedWanderer? otherWanderer;

  final Vector2 position;

  final Vector2 velocity;

  final Size worldSize;

  PairedWanderer({
    required this.position,
    required this.velocity,
    required this.worldSize,
  });

  void update(double dt) {
    position.addScaled(velocity, dt);
    if (otherWanderer != null) {
      position.addScaled(otherWanderer!.velocity, dt * 0.25);
    }

    if (position.x < 0 && velocity.x < 0) {
      velocity.x = -velocity.x;
    } else if (position.x > worldSize.width && velocity.x > 0) {
      velocity.x = -velocity.x;
    }
    if (position.y < 0 && velocity.y < 0) {
      velocity.y = -velocity.y;
    } else if (position.y > worldSize.height && velocity.y > 0) {
      velocity.y = -velocity.y;
    }
  }
}

class PairedWandererWidget extends StatefulWidget {
  final PairedWanderer wanderer;

  const PairedWandererWidget({required this.wanderer, super.key});

  @override
  State<PairedWandererWidget> createState() => _PairedWandererWidgetState();
}

class _PairedWandererWidgetState extends State<PairedWandererWidget>
    with SingleTickerProviderStateMixin {
  late Ticker _ticker;

  Duration _lastElapsed = Duration.zero;

  @override
  void initState() {
    super.initState();
    _ticker = createTicker(_onTick);
    _ticker.start();
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: widget.wanderer.position.x - 128 / 4,
      top: widget.wanderer.position.y - 128 / 4,
      child: const SizedBox(
        width: 8,
        height: 8,
        child: Placeholder(),
      ),
    );
  }

  void _onTick(Duration elapsed) {
    var dt = (elapsed - _lastElapsed).inMicroseconds / 1000000;
    dt = min(dt, 1 / 60);
    widget.wanderer.update(dt);
    _lastElapsed = elapsed;
    setState(() {});
  }
}
```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

The code sample is compatible with Dartpad.

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
flutter doctor -v
[✓] Flutter (Channel stable, 3.24.0, on macOS 14.4.1 23E224 darwin-arm64,
    locale en-US)
    • Flutter version 3.24.0 on channel stable at
      /Users/filiph/fvm/versions/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 80c2e84975 (3 weeks ago), 2024-07-30 23:06:49 +0700
    • Engine revision b8800d88be
    • Dart version 3.5.0
    • DevTools version 2.37.2

[✓] Android toolchain - develop for Android devices (Android SDK version
    35.0.0)
    • Android SDK at /Users/filiph/Library/Android/sdk
    • Platform android-35, build-tools 35.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.7+0-17.0.7b1000.6-10550314)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.3.2)
    • IntelliJ at /Users/filiph/Applications/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 78.4.2
    • Dart plugin version 233.15271

[✓] VS Code (version 1.87.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.84.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64
      • macOS 14.4.1 23E224 darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin
      • macOS 14.4.1 23E224 darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript
      • Google Chrome 127.0.6533.120

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by huycozy** [comment]
From my end, I observed the Memory in Chrome's Devtools and noticed that: when I wait long enough, the memory is raised to 7xx-8xx MB but it could be dropped to 4xx-5xx MB afterward and so on (up and down sequentially)

![Screenshot 2024-08-20 at 15 22 32](https://github.com/user-attachments/assets/665617ef-718f-431b-a41d-9ea29ba2e2a8)

Flutter version: `Channel master, 3.25.0-1.0.pre.27`

**Comment by parancode** [comment]
It seems like every time a widget is rendered, a memory rick occurs.

In some cases, my product has more than 3GB of memory (which automatically shuts down on Chrome if it exceeds that). The canvaskit uses more memory, so the memory won't trigger the flutter engine faster on devices with built-in graphics.

**Comment by filiph** [comment]
> From my end, I observed the Memory in Chrome's Devtools and noticed that: when I wait long enough, the memory is raised to 7xx-8xx MB but it could be dropped to 4xx-5xx MB afterward and so on (up and down sequentially)

Confirmed. I got that behavior, too, eventually, but in my case it rose to 2GB before dropping to 7xx MB and raising again.

**Comment by mraleph** [comment]
I think this is another manifestation of https://issues.chromium.org/issues/340777103, there is no real leak - but the reclamation of objects is so delayed that it looks like a leak.  

`_PlaceholderPainter.paint` creates a bunch of native objects wrapped into Dart objects (e.g. `Paint` and `Path`). Lifetime of native objects and Dart objects are tied together via `FinalizationRegistry` but this means that due to V8 behavior these objects have prolonged lifetime.

cc @yjbanov @eyebrowsoffire

**Comment by mraleph** [comment]
You can confirm that this is caused by `Placeholder` widget by doing quick and dirty patch which caches `Path` and `Paint` objects:  

<details>

```diff
diff --git a/packages/flutter/lib/src/widgets/placeholder.dart b/packages/flutter/lib/src/widgets/placeholder.dart
index 1fdd6e7e27..82156d5561 100644
--- a/packages/flutter/lib/src/widgets/placeholder.dart
+++ b/packages/flutter/lib/src/widgets/placeholder.dart
@@ -8,7 +8,7 @@ import 'basic.dart';
 import 'framework.dart';
 
 class _PlaceholderPainter extends CustomPainter {
-  const _PlaceholderPainter({
+  _PlaceholderPainter({
     required this.color,
     required this.strokeWidth,
   });
@@ -16,24 +16,33 @@ class _PlaceholderPainter extends CustomPainter {
   final Color color;
   final double strokeWidth;
 
-  @override
-  void paint(Canvas canvas, Size size) {
-    final Paint paint = Paint()
-      ..color = color
-      ..style = PaintingStyle.stroke
-      ..strokeWidth = strokeWidth;
-    final Rect rect = Offset.zero & size;
-    final Path path = Path()
+  late final Paint _paint = Paint()
+    ..color = color
+    ..style = PaintingStyle.stroke
+    ..strokeWidth = strokeWidth;
+  Size _size = Size.zero;
+  late Path _path = _computePath(_size);
+
+  static Path _computePath(Size size) {
+    final rect = Offset.zero & size;
+    return Path()
       ..addRect(rect)
       ..addPolygon(<Offset>[rect.topRight, rect.bottomLeft], false)
       ..addPolygon(<Offset>[rect.topLeft, rect.bottomRight], false);
-    canvas.drawPath(path, paint);
+  }
+
+  @override
+  void paint(Canvas canvas, Size size) {
+    if (_size != size) {
+      _size = size;
+      _path = _computePath(size);
+    }
+    canvas.drawPath(_path, _paint);
   }
 
   @override
   bool shouldRepaint(_PlaceholderPainter oldPainter) {
-    return oldPainter.color != color
-        || oldPainter.strokeWidth != strokeWidth;
+    return oldPainter.color != color || oldPainter.strokeWidth != strokeWidth;
   }
 
   @override
@@ -53,14 +62,13 @@ class _PlaceholderPainter extends CustomPainter {
 /// {@youtube 560 315 https://www.youtube.com/watch?v=LPe56fezmoo}
 class Placeholder extends StatelessWidget {
   /// Creates a widget which draws a box.
-  const Placeholder({
-    super.key,
-    this.color = const Color(0xFF455A64), // Blue Grey 700
-    this.strokeWidth = 2.0,
-    this.fallbackWidth = 400.0,
-    this.fallbackHeight = 400.0,
-    this.child
-  });
+  const Placeholder(
+      {super.key,
+      this.color = const Color(0xFF455A64), // Blue Grey 700
+      this.strokeWidth = 2.0,
+      this.fallbackWidth = 400.0,
+      this.fallbackHeight = 400.0,
+      this.child});
 
   /// The color to draw the placeholder box.
   final Color color;
@@ -107,9 +115,13 @@ class Placeholder extends StatelessWidget {
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder properties) {
     super.debugFillProperties(properties);
-    properties.add(ColorProperty('color', color, defaultValue: const Color(0xFF455A64)));
-    properties.add(DoubleProperty('strokeWidth', strokeWidth, defaultValue: 2.0));
-    properties.add(DoubleProperty('fallbackWidth', fallbackWidth, defaultValue: 400.0));
-    properties.add(DoubleProperty('fallbackHeight', fallbackHeight, defaultValue: 400.0));
+    properties.add(
+        ColorProperty('color', color, defaultValue: const Color(0xFF455A64)));
+    properties
+        .add(DoubleProperty('strokeWidth', strokeWidth, defaultValue: 2.0));
+    properties.add(
+        DoubleProperty('fallbackWidth', fallbackWidth, defaultValue: 400.0));
+    properties.add(
+        DoubleProperty('fallbackHeight', fallbackHeight, defaultValue: 400.0));
   }
 }
```

</details>

**Comment by yjbanov** [comment]
@mraleph @leafpetersen 

This is a good example of untimely resource collection that we discussed the other day. The heuristics of the GC can be felt in real ways, even in otherwise perfect Dart language implementations. Games in particular are very time sensitive. Finding a robust method of timely resource reclamation approach could help a lot (e.g. ref-counted disposable objects). For example, in this function the language could somehow know that `paint` and `path` variables can be safely disposed of before the function returns.

```dart
  @override
  void paint(Canvas canvas, Size size) {
    final Paint paint = Paint()
      ..color = color
      ..style = PaintingStyle.stroke
      ..strokeWidth = strokeWidth;
    final Rect rect = Offset.zero & size;
    final Path path = Path()
      ..addRect(rect)
      ..addPolygon(<Offset>[rect.topRight, rect.bottomLeft], false)
      ..addPolygon(<Offset>[rect.topLeft, rect.bottomRight], false);
    canvas.drawPath(path, paint);
  }
```

Ideally, the solution would not require the developer to rewrite this code. For example, the `Paint` and `Path` classes could be made `implements Disposable` where `Disposable` is a `dart:core` type known to the compiler. The `Canvas.drawPath` method would be annotated that it `borrows` its arguments (to use the Rust parlance), signalling that `drawPath` doesn't share these variable with anything else. Then the compiler would automatically call `path.dispose()` and `paint.dispose()` before returning.

In the meantime, we could just add `dispose()` methods to these types, and developers would cherry-pick places to call them. That style is prone to all the C-style resource management errors (dangling pointers, double-frees, etc), but maybe it's _something_?

**Comment by mraleph** [comment]
I think finding "eager reclamation" semantics which both fits Dart language and does not require massive type system investments is rather challenging, e.g. it is easy to try promising something like _`Disposable` objects have their `dispose` called automatically_, but it just does not work because you need to know which particular reference actually "owns" the object you are going to dispose. 

```
class A implements Disposable {
  final B b;
  A(this.b);

  @override
  void dispose() {
  }
}

class B implements Disposable {
  @override
  void dispose() {
  }  
}

void foo() {
  final b = B();
  final a = A(b);
}
```

`foo` should call `a.dispose(...)` but `b.dispose()` should likely occur inside `A.dispose`. But this requires ownership to be incorporated into the type-system. 

I have done some experiments with [interprocedural escape analysis](https://github.com/dart-lang/sdk/compare/main...mraleph:sdk:escape-analysis) - it can correctly compute that for many temporary objects like `Paint` that they don't outlive the frame in which they are allocated - this information can be used to place eager `dispose` calls in these frames - but I don't like that this is just an optimization which might or might not succeed rather than a language feature which guarantees certain properties.

**Comment by yjbanov** [comment]
Update: `Paint` objects are no longer stored in the C++ heap. However, the repro still runs out of memory eventually. I'll be looking at paths next when I get the chance, although maybe it's more than just paints and paths.

**Comment by polina-c** [comment]
TLDR - i could repro and tweak the code to fix the leak

Here is my observations:

1. Yes, I can repro it and get to ~1500 MB, and it did not stop
2. [Checked with leak tracker](https://github.com/flutter/flutter/pull/158964) - yes, all disposables are disposed on mac (i understand it is not issue on other platforms, but it could be not disposed objects do not result in leaks)
3. I removed SingleTickerProviderStateMixin and AnimationController from _MemoryPressureWidgetState, as they are noop. It did not fix the issue.
4. I updated _PairedWandererWidgetState to simply use Timer instead of SingleTickerProviderStateMixin. It did not change how it look, but **fixed the leak** - memory climbed maximum to 491 MB.

<details>
 <summary>Updated _PairedWandererWidgetState:</summary>
```
class _PairedWandererWidgetState extends State<PairedWandererWidget>
//with SingleTickerProviderStateMixin
{
  late Duration _lastElapsed = Duration.zero;
  late Timer _ticker;
  final startTime = DateTime.now();

  @override
  void initState() {
    super.initState();

    _ticker = Timer.periodic(const Duration(milliseconds: 16), _onTick);
  }

  @override
  void dispose() {
    _ticker.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: widget.wanderer.position.x - 128 / 4,
      top: widget.wanderer.position.y - 128 / 4,
      child: const SizedBox(
        width: 8,
        height: 8,
        child: Placeholder(),
      ),
    );
  }

  void _onTick(Timer timer) {
    final elapsed = DateTime.now().difference(startTime);
    var dt = (elapsed - _lastElapsed).inMicroseconds / 1000000;
    dt = min(dt, 1 / 60);
    widget.wanderer.update(dt);
    _lastElapsed = elapsed;
    setState(() {});
  }
}
```
</details> 

What else we can do:
1. Try to deep dive to SingleTickerProviderStateMixin, to check what is wrong here on web
2. I can hack leak tracker to enable it for web for this test and check all disposables are disposed

Thoughts?

**Comment by yjbanov** [comment]
@polina-c  Thanks for narrowing it down to SingleTickerProviderStateMixin! This is super-useful. As of right now I cannot think of anything that would be web-specific in it. However, it's a great place to start.

**Comment by Yerdna1** [comment]
ok, I have ListView.Builder, with lot of checkboxes, but on this page I am using class _UserTemplateDetailEditState extends State<UserTemplateDetailEdit> with SingleTickerProviderStateMixin. In Firefox it is freezing ,or memory leaks to 3GB with only Scrolling Up and Scrolling Down, so looks like, that issue is really in Flutter 3.24.5 in SingleTickerProviderStateMixin for flutter_web

**Comment by polina-c** [comment]
I updated leak tracker to monitor web (will post pull request later, after figuring out this issue). It did not find not any disposed disposables when running `flutter test --platform chrome` for the test that pumps the app from the example and then delay.

I executed example with `flutter run -d chrome --profile --dart-define=FLUTTER_WEB_ENABLE_INSTRUMENTATION=true`. Then, when I closed Chrome, the test process did not finish. How can I see the screen shot you from [this comment](https://github.com/flutter/flutter/issues/153678#issuecomment-2477164203)? 

@yjbanov

**Comment by yjbanov** [comment]
Tests use DDC, and I don't know if DDC supports environment variables. I just `flutter run` the app and look at the JavaScript console.

**Comment by jezell** [comment]
We may also be encountering this. We've been seeing something occasionally where chrome memory usage can be seen to rise hundreds of megs per second until the OS itself runs out of memory. We've seen Chrome use 20GB+ of memory when it's happening until the OS itself complains that it's out of memory. Sounds suspiciously similar. Definitely has felt like something was leaking in v8 / chrome itself under some conditions where it can't keep up with the engine because there's no way Flutter could use that amount of memory even if it wanted to.

**Comment by jason-simmons** [comment]
Looked into the picture leaks reported in https://github.com/flutter/flutter/issues/153678#issuecomment-2477164203.

Those pictures appear to be coming from the rendering canvases used by `HtmlViewEmbedder.submitFrame`.

If I add a `renderPicture.dispose()` call after the `rasterizer.rasterizeToCanvas` call in the `renderCanvas` loop, then I do not see the "Picture Leaked" counter incrementing.

However, disposing these pictures did not have a noticeable effect on this app's overall memory usage.

**Comment by polina-c** [comment]
It may be that SingleTickerProviderStateMixin is doing something that holds objects from being GCed.

**Comment by polina-c** [comment]
Good news: it is easier to detect not garbage collected objects, when everything is disposed. Disposal means an object is not needed any more and is expected to be GCed nearest GC cycle.

**Comment by polina-c** [comment]
@jason-simmons, I tried to repro the issue with the latest version of Flutter and the memory did not climb to more than 200 MB.

But you stated that "disposing these pictures did not have a noticeable effect on this app's overall memory usage". Does the issue still repro for you?

**Comment by jason-simmons** [comment]
I'm still seeing this issue.  On my Linux setup, the VmRSS of the Chrome renderer process grows to 1.4GB when running the example app in release mode.

I also see this happen if I apply the patches from https://github.com/flutter/flutter/issues/153678#issuecomment-2477623091 and remove the usage of `SingleTickerProviderStateMixin`.

**Comment by polina-c** [comment]
Thanks. How exactly you check VmRSS?

**Comment by polina-c** [comment]
@jason-simmons , I do not have Fluter configured on linux. May I ask you to check one thing?

Can you add `ImageCache.clear();` to `dispose` of _PairedWandererWidgetState and check if the issue repro?

**Comment by jason-simmons** [comment]
`_PairedWandererWidgetState.dispose()` is not called while the animation is running.

I did not notice any effect on Chrome renderer process memory usage from adding an `imageCache.clear()` call to `_PairedWandererWidgetState. _onTick`.

**Comment by yjbanov** [comment]
Dropped to P2 as there's no additional active work happening right now (although https://github.com/flutter/flutter/pull/168384 might have helped here too). We may bump it back up again, if we can discover some clear signals we can act on.

**Comment by mraleph** [comment]
@yjbanov @eyebrowsoffire do we have some good example that illustrates the type of problem we are experiencing due to https://issues.chromium.org/issues/340777103? I have tried the original repro from this issue - and while I can still see a slow creep upwards which only gets resolved by major GC - it is not as drastic as previous issue. I assume that issue is not that obvious anymore because we change implementation of `Paint` objects.

cc @mlippautz

**Comment by eyebrowsoffire** [comment]
@mraleph We have done a number of changes recently to reduce our reliance on the GC to clean up native objects:
* As you mentioned, a while ago @yjbanov changed the `Paint` object to not rely on the GC anymore. It is created just-in-time when passing to native APIs and immediately released on the dart side, so the object is only managed through native shared pointer stuff.
* I did something similar for the `Path` object (both skwasm and CanvasKit), where we explicitly free path objects at the end of the frame, and if they are reused across frames we rebuild them: https://github.com/flutter/flutter/pull/170303
* I did a number of other changes to make text-related objects' memory lifecycle independent of GC in skwasm. This doesn't make any changes to the canvaskit build though: https://github.com/flutter/flutter/pull/172187

As far as I know, at least on the skwasm side the only thing we're still relying on the GC for is the `Vertices` object, which is a bit tricky to fix and a relatively uncommon use case.
